# **Write a program to implement k-Nearest Neighbor algorithm to classify the weather dataset. Print confusion matrix. (k=5)**

### **Import statements**

In [ ]:
import pandas as pd
from sklearn import metrics
from sklearn.model_selection import train_test_split 
from sklearn import preprocessing

### **Reading the csv file**

In [ ]:
data = pd.read_csv('weather.csv')
data.head()

MinTemp  MaxTemp  Rainfall  ...  RainToday  RISK_MM RainTomorrow
0      8.0     24.3       0.0  ...         No      3.6          Yes
1     14.0     26.9       3.6  ...        Yes      3.6          Yes
2     13.7     23.4       3.6  ...        Yes     39.8          Yes
3     13.3     15.5      39.8  ...        Yes      2.8          Yes
4      7.6     16.1       2.8  ...        Yes      0.0           No

[5 rows x 22 columns]

### **Checking for null values**

In [ ]:
data.isnull().sum()

MinTemp           0
MaxTemp           0
Rainfall          0
Evaporation       0
Sunshine          3
WindGustDir       3
WindGustSpeed     2
WindDir9am       31
WindDir3pm        1
WindSpeed9am      7
WindSpeed3pm      0
Humidity9am       0
Humidity3pm       0
Pressure9am       0
Pressure3pm       0
Cloud9am          0
Cloud3pm          0
Temp9am           0
Temp3pm           0
RainToday         0
RISK_MM           0
RainTomorrow      0
dtype: int64

### **Shape of dataframe**

In [ ]:
data.shape

(366, 22)

### **Dropping all null and missing values**

In [ ]:
data = data.dropna()
data

MinTemp  MaxTemp  Rainfall  ...  RainToday  RISK_MM RainTomorrow
0        8.0     24.3       0.0  ...         No      3.6          Yes
1       14.0     26.9       3.6  ...        Yes      3.6          Yes
2       13.7     23.4       3.6  ...        Yes     39.8          Yes
3       13.3     15.5      39.8  ...        Yes      2.8          Yes
4        7.6     16.1       2.8  ...        Yes      0.0           No
..       ...      ...       ...  ...        ...      ...          ...
361      9.0     30.7       0.0  ...         No      0.0           No
362      7.1     28.4       0.0  ...         No      0.0           No
363     12.5     19.9       0.0  ...         No      0.0           No
364     12.5     26.9       0.0  ...         No      0.0           No
365     12.3     30.2       0.0  ...         No      0.0           No

[328 rows x 22 columns]

(328, 22) is the shape after removing all null values which does not affect our dataset much. Earlier, it was (366, 22)

### **Encoding the labels of the output to from (yes, no) to (1, 0)**

In [ ]:
data.replace('Yes', '1', inplace=True)
data.replace('No', '0', inplace=True)
data

MinTemp  MaxTemp  Rainfall  ...  RainToday  RISK_MM RainTomorrow
0        8.0     24.3       0.0  ...          0      3.6            1
1       14.0     26.9       3.6  ...          1      3.6            1
2       13.7     23.4       3.6  ...          1     39.8            1
3       13.3     15.5      39.8  ...          1      2.8            1
4        7.6     16.1       2.8  ...          1      0.0            0
..       ...      ...       ...  ...        ...      ...          ...
361      9.0     30.7       0.0  ...          0      0.0            0
362      7.1     28.4       0.0  ...          0      0.0            0
363     12.5     19.9       0.0  ...          0      0.0            0
364     12.5     26.9       0.0  ...          0      0.0            0
365     12.3     30.2       0.0  ...          0      0.0            0

[328 rows x 22 columns]

### **Encoding the columns with string elements to numerical (int64) values**

In [ ]:
unique_dir = list(pd.unique(data['WindGustDir']))
print(unique_dir)

['NW', 'ENE', 'SSE', 'SE', 'E', 'S', 'N', 'WNW', 'ESE', 'NE', 'NNE', 'NNW', 'SW', 'W', 'WSW', 'SSW']


In [ ]:
data = data.replace({'WindGustDir': {'NW': 1, 'ENE': 2, 'SSE': 3, 'SE': 4,'E': 5, 'S': 6,'N': 7, 'WNW': 8,'ESE': 9, 'NE': 10,'NNE': 11, 'NNW': 12,'SW': 13, 'W': 14,'WSW': 15, 'SSW': 16}}) 
data

MinTemp  MaxTemp  Rainfall  ...  RainToday  RISK_MM  RainTomorrow
0        8.0     24.3       0.0  ...          0      3.6             1
1       14.0     26.9       3.6  ...          1      3.6             1
2       13.7     23.4       3.6  ...          1     39.8             1
3       13.3     15.5      39.8  ...          1      2.8             1
4        7.6     16.1       2.8  ...          1      0.0             0
..       ...      ...       ...  ...        ...      ...           ...
361      9.0     30.7       0.0  ...          0      0.0             0
362      7.1     28.4       0.0  ...          0      0.0             0
363     12.5     19.9       0.0  ...          0      0.0             0
364     12.5     26.9       0.0  ...          0      0.0             0
365     12.3     30.2       0.0  ...          0      0.0             0

[328 rows x 22 columns]

### **Getting all numerical values**

In [ ]:
data = data.replace({'WindDir3pm': {'NW': 1, 'ENE': 2, 'SSE': 3, 'SE': 4,'E': 5, 'S': 6,'N': 7, 'WNW': 8,'ESE': 9, 'NE': 10,'NNE': 11, 'NNW': 12,'SW': 13, 'W': 14,'WSW': 15, 'SSW': 16}}) 
data = data.replace({'WindDir9am': {'NW': 1, 'ENE': 2, 'SSE': 3, 'SE': 4,'E': 5, 'S': 6,'N': 7, 'WNW': 8,'ESE': 9, 'NE': 10,'NNE': 11, 'NNW': 12,'SW': 13, 'W': 14,'WSW': 15, 'SSW': 16}}) 
data

MinTemp  MaxTemp  Rainfall  ...  RainToday  RISK_MM  RainTomorrow
0        8.0     24.3       0.0  ...          0      3.6             1
1       14.0     26.9       3.6  ...          1      3.6             1
2       13.7     23.4       3.6  ...          1     39.8             1
3       13.3     15.5      39.8  ...          1      2.8             1
4        7.6     16.1       2.8  ...          1      0.0             0
..       ...      ...       ...  ...        ...      ...           ...
361      9.0     30.7       0.0  ...          0      0.0             0
362      7.1     28.4       0.0  ...          0      0.0             0
363     12.5     19.9       0.0  ...          0      0.0             0
364     12.5     26.9       0.0  ...          0      0.0             0
365     12.3     30.2       0.0  ...          0      0.0             0

[328 rows x 22 columns]

### **Checking the data types, if they are all numerical**

In [ ]:
data.dtypes

MinTemp          float64
MaxTemp          float64
Rainfall         float64
Evaporation      float64
Sunshine         float64
WindGustDir        int64
WindGustSpeed    float64
WindDir9am         int64
WindDir3pm         int64
WindSpeed9am     float64
WindSpeed3pm       int64
Humidity9am        int64
Humidity3pm        int64
Pressure9am      float64
Pressure3pm      float64
Cloud9am           int64
Cloud3pm           int64
Temp9am          float64
Temp3pm          float64
RainToday         object
RISK_MM          float64
RainTomorrow      object
dtype: object

### **Converting object data types to integer type**

In [ ]:
data['RainToday'] = data['RainToday'].astype(int)
data['RainTomorrow'] = data['RainTomorrow'].astype(int)
data

MinTemp  MaxTemp  Rainfall  ...  RainToday  RISK_MM  RainTomorrow
0        8.0     24.3       0.0  ...          0      3.6             1
1       14.0     26.9       3.6  ...          1      3.6             1
2       13.7     23.4       3.6  ...          1     39.8             1
3       13.3     15.5      39.8  ...          1      2.8             1
4        7.6     16.1       2.8  ...          1      0.0             0
..       ...      ...       ...  ...        ...      ...           ...
361      9.0     30.7       0.0  ...          0      0.0             0
362      7.1     28.4       0.0  ...          0      0.0             0
363     12.5     19.9       0.0  ...          0      0.0             0
364     12.5     26.9       0.0  ...          0      0.0             0
365     12.3     30.2       0.0  ...          0      0.0             0

[328 rows x 22 columns]

### **Final check of all data types**

In [ ]:
data.dtypes

MinTemp          float64
MaxTemp          float64
Rainfall         float64
Evaporation      float64
Sunshine         float64
WindGustDir        int64
WindGustSpeed    float64
WindDir9am         int64
WindDir3pm         int64
WindSpeed9am     float64
WindSpeed3pm       int64
Humidity9am        int64
Humidity3pm        int64
Pressure9am      float64
Pressure3pm      float64
Cloud9am           int64
Cloud3pm           int64
Temp9am          float64
Temp3pm          float64
RainToday          int64
RISK_MM          float64
RainTomorrow       int64
dtype: object

### **Defining x (input features) and y (output variable)**

where all the columns are considered for KNN Algorithm and y = RainTomorrow will predict if there will be rainfall tomorrow

In [ ]:
x = data.drop(['RainTomorrow'], axis=1)
y = data['RainTomorrow']
y.head()

0    1
1    1
2    1
3    1
4    0
Name: RainTomorrow, dtype: int64

### **Implementing the KNNClassifier Model for k=5**

In [ ]:
from sklearn.neighbors import KNeighborsClassifier
model = KNeighborsClassifier(n_neighbors=5)

### **Splitting the data**

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.20)
model.fit(x_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
                     metric_params=None, n_jobs=None, n_neighbors=5, p=2,
                     weights='uniform')

### **Predicting the output**

In [ ]:
MinTemp = 8.0
MaxTemp	= 24.3
Rainfall = 0.0
Evaporation	= 3.4
Sunshine = 	6.3
WindGustDir = 1
WindGustSpeed = 30.0	
WindDir9am = 13	
WindDir3pm = 1
WindSpeed9am = 	6.0
WindSpeed3pm = 	20
Humidity9am	= 68
Humidity3pm	= 29
Pressure9am	= 1019.7
Pressure3pm	= 1015.0
Cloud9am = 7
Cloud3pm = 7
Temp9am = 14.4
Temp3pm = 23.6
RainToday = 0	
RISK_MM = 3.6
print(model.predict([[MinTemp, MaxTemp, Rainfall, Evaporation, Sunshine, WindGustDir, WindGustSpeed, WindDir9am, WindDir3pm, WindSpeed9am, WindSpeed3pm, Humidity9am, Humidity3pm, Pressure9am, Pressure3pm, Cloud9am, Cloud3pm, Temp9am, Temp3pm, RainToday, RISK_MM]]))

[0]


### **Accuracy of the model**

In [ ]:
y_pred = model.predict(x_test)
score = metrics.accuracy_score(y_test,y_pred)
print(score*100,'%')

83.33333333333334 %


### **Confusion Matrix**

In [ ]:
from sklearn.metrics import confusion_matrix

print('Confusion Matrix:')
print(confusion_matrix(y_test, y_pred))

Confusion Matrix:
[[53  1]
 [10  2]]
